integration/internal/dctest/cmd.go

```diff
@@ -0,0 +1,96 @@
+package dctest
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/concourse/concourse/integration/internal/cmdtest"
+	"github.com/concourse/concourse/integration/internal/ypath"
+	"github.com/stretchr/testify/require"
+)
+
+type Cmd struct {
+	cmdtest.Cmd
+}
+
+func Init(t *testing.T, composeFile string, overrides ...string) Cmd {
+	name := filepath.Base(t.Name())
+
+	files := append([]string{composeFile}, overrides...)
+
+	dc := cmdtest.Cmd{
+		Path: "docker-compose",
+		Env: []string{
+			"COMPOSE_FILE=" + strings.Join(files, ":"),
+			"COMPOSE_PROJECT_NAME=" + name,
+		},
+	}
+
+	for _, env := range os.Environ() {
+		if strings.HasPrefix(env, "TEST_") {
+			dc = dc.WithEnv(env)
+		}
+	}
+
+	// clean up docker-compose when the test finishes
+	cleanupOnce(t, func() {
+		if t.Failed() {
+			err := os.MkdirAll("logs", os.ModePerm)
+			if err == nil {
+				logFile, err := os.Create("logs/" + name + ".log")
+				if err == nil {
+					dc.Silence().OutputTo(logFile).Run("logs", "--no-color")
+					logFile.Close()
+				}
+			}
+		}
+
+		dc.Run("kill")
+		dc.Run("down")
+	})
+
+	return Cmd{
+		Cmd: dc,
+	}
+}
+
+func InitDynamic(t *testing.T, doc *ypath.Document) Cmd {
+	name := filepath.Base(t.Name())
+	fileName := fmt.Sprintf(".docker-compose.%s.yml", name)
+
+	err := ioutil.WriteFile(fileName, doc.Bytes(), os.ModePerm)
+	require.NoError(t, err)
+
+	cleanupOnce(t, func() {
+		os.Remove(fileName)
+	})
+
+	return Init(t, fileName)
+}
+
+func (cmd Cmd) Addr(container string, port int) (string, error) {
+	out, err := cmd.Output("port", container, strconv.Itoa(port))
+	if err != nil {
+		return "", err
+	}
+
+	return strings.TrimRight(strings.Replace(out, "0.0.0.0", "127.0.0.1", 1), "\n"), nil
+}
+
+var onces = new(sync.Map)
+
+func cleanupOnce(t *testing.T, cleanup func()) {
```

Comment:
When could the `Cleanup` funcs get called more than once?

Reply:
Ignore me, I think I see it now - we spin up multiple `dctest.Cmd`'s, but they actually point to the same deployment, so we just want to `docker-compose down` once?

I wonder if we'll ever have a single test that needs two different deployments? Not sure why we would need to do that, rather than just adding overrides or using the `ypath` package to create a more complicated `docker-compose` deployment to suit our needs, though

Reply:
If a test calls `dctest.Init` twice - e.g. the upgrade/downgrade tests. It won't call each function twice, but it would call both functions, and in this case we should only need to run one `docker-compose` cleanup, and it should be whichever one was added last, which works since `t.Cleanup` is called in reverse order.